Sources/Atomic.swift

```diff
@@ -163,18 +186,30 @@ public final class Atomic<Value>: AtomicProtocol {
 		lock = PosixThreadMutex()
 	}
 
+	deinit {
+		lock.deinitialize()
+	}
+
 	/// Atomically modifies the variable.
 	///
+	/// - note: Throwing in `action` does not revert any changes made.
+	///
 	/// - parameters:
 	///   - action: A closure that takes the current value.
 	///
 	/// - returns: The result of the action.
 	@discardableResult
+	@inline(__always)
 	public func modify<Result>(_ action: (inout Value) throws -> Result) rethrows -> Result {
 		lock.lock()
-		defer { lock.unlock() }
-
-		return try action(&_value)
+		do {
+			let value = try action(&_value)
+			lock.unlock()
+			return value
+		} catch {
+			lock.unlock()
+			throw error
+		}
```

Comment:
How likely are we to change this though? It are pieces that perhaps would never be altered otherwise.